upstream.go

```diff
@@ -508,6 +548,88 @@ func (uc *upstreamConn) handleMessage(msg *irc.Message) error {
 				Params:  params,
 			})
 		})
+	case "MODE":
+		var name, modeStr string
+		if err := parseMessageParams(msg, &name, &modeStr); err != nil {
+			return err
+		}
+
+		if !uc.isChannel(name) { // user mode change
+			if name != uc.nick {
+				return fmt.Errorf("received MODE message for unknown nick %q", name)
+			}
+			return uc.modes.Apply(modeStr)
+			// TODO: notify downstreams about user mode change?
+		} else { // channel mode change
+			ch, err := uc.getChannel(name)
+			if err != nil {
+				return err
+			}
+
+			if ch.modes != nil {
+				if err := ch.modes.Apply(uc.availableChannelModes, modeStr, msg.Params[2:]...); err != nil {
+					return err
+				}
+			}
+
+			uc.forEachDownstream(func(dc *downstreamConn) {
+				params := []string{dc.marshalChannel(uc, name), modeStr}
+				params = append(params, msg.Params[2:]...)
+
+				dc.SendMessage(&irc.Message{
+					Prefix:  dc.marshalUserPrefix(uc, msg.Prefix),
+					Command: "MODE",
+					Params:  params,
+				})
+			})
+		}
+	case irc.RPL_UMODEIS:
+		if err := parseMessageParams(msg, nil); err != nil {
+			return err
+		}
+		modeStr := ""
+		if len(msg.Params) > 1 {
+			modeStr = msg.Params[1]
+		}
+
+		uc.modes = ""
+		if err := uc.modes.Apply(modeStr); err != nil {
+			return err
+		}
+	case irc.RPL_CHANNELMODEIS:
+		var channel string
+		if err := parseMessageParams(msg, nil, &channel); err != nil {
+			return err
+		}
+		modeStr := ""
+		if len(msg.Params) > 2 {
+			modeStr = msg.Params[2]
+		}
+
+		ch, err := uc.getChannel(channel)
+		if err != nil {
+			return err
+		}
+
+		firstMode := ch.modes == nil
+		ch.modes = make(map[byte]string)
+		if err := ch.modes.Apply(uc.availableChannelModes, modeStr, msg.Params[3:]...); err != nil {
+			return err
+		}
+		if firstMode {
```

Comment:
We should probably send `RPL_CHANNELMODEIS` to downstream connections unconditionally. Otherwise clients may become out-of-sync.